resolver/graph.go

```diff
@@ -337,23 +335,15 @@ func CreateMessageRuleDependencyGraph(ctx *context, baseMsg *Message, rule *Mess
 		}
 	}
 
-	nodeMap := make(map[*MessageRuleDependencyGraphNode]struct{})
-	for _, node := range msgToNode {
-		nodeMap[node] = struct{}{}
-	}
-	for _, node := range nameToNode {
-		nodeMap[node] = struct{}{}
+	allNodes := make([]*MessageRuleDependencyGraphNode, 0, len(msgToNodes))
```

Comment:
Would it be computationally cheaper if we only sort the roots? Currently, it sorts `allNodes` but IIUC, we only want sorted roots, so what if we first get roots and sort them instead?

```
	var roots []*MessageRuleDependencyGraphNode
	for _, nodes := range msgToNodes {
		for _, node := range nodes {
			if len(node.Parent) == 0 {
				roots = append(roots, node)
			}
		}
	}
	sort.Slice(roots, func(i, j int) bool {
		return roots[i].FQDN() < roots[j].FQDN()
	})
```